### DISCUSSION TITLE
32-bit

### DISCUSSION THREAD

**Discussion by psidome** [NTNU-IHB/PythonFMU#143] [Q&A]
Hi,
Is it possible to create a .dll-file for 32-bit windows? I've tried a lot of things, but i just keep getting linux64 and win64 folders in the fmu. I want to import the FMU into LabVIEW, which only suppports 32-bit dlls from an FMU.

**Comment by markaren** [comment]
It is possible if you build the native library locally. 
https://github.com/NTNU-IHB/PythonFMU/tree/master/pythonfmu/pythonfmu-export
Just make sure you use a 32-bit compiler. Take the generated .dll and drop it into the FMU archive.

**Reply by psidome** [reply]
Let's say  I want to build the "demoslave.py" in the examples folder. Do I need to copy it somewhere or edit the Cmakelists.txt or just include the path in the command line? If I run the commands from within "build_win.sh" something is compiled, but I don't know what.

**Comment by markaren** [comment]
`build_win.sh` is configured for 64-bit. I guess you could just remove `x64` to compile for 32-bit.
It should create a win32 folder in pythonfmu/resources I guess.

Just build the FMU using the command line tool `pythonfmu build -f "script.py"` and drop the .dll folder into binaries.

And naturally, your Python installation must also be 32-bit...

**Comment by Daiiiy** [comment]
Hi,
I've tried to use following command to build 32-bit fmu.   (Anaconda 32-bit、python3.9 32-bit、VS2019、pyfmi)
!!!

- mkdir tmp-build
- cd tmp-build
- cmake ../pythonfmu/pythonfmu-export -DCMAKE_BUILD_TYPE=Release -A Win32
- cmake --build . --config Release
- cd ..
- python -m pip install .

!!!
Then I got 32-bit pythonfmu,it is possible to pack 32-bit fmu,when I load this fmu using [pyfmi](https://pypi.org/project/PyFMI/),python.exe exits directly.
I don't know why this happens, have you ever successfully compiled a 32-bit fmu case?

**Reply by markaren** [reply]
I would assume it should work. Is pyfmi 32 bit? Does it work with FMPy? Also, I am curious why you need to build 32-bit at all?

**Reply by Daiiiy** [reply]
Thank you for your reply，
my pyfmi is 32 bit,and i also tried using FMPy,it doesn't work either.The FMPy GUI freezes when simulating.
By the way,I am developing an FMI interface for my fortran program (32-bit) which is hard to change to win64. 
I want to be able to import python code into my program as fmu.

**Comment by havifloja** [comment]
I can confirm the issues with 32bit python reported above. Debugging reveals that the `PySlaveInstance::initialize` function freezes.
We could resolve this issue by changing `pythonfmu\pythonfmu-export\cpp\PySlaveInstance.cpp` (version 0.6.2) as follows:

```
PyObject* kwargs = Py_BuildValue("{ss,ss,sL,si}",
        "instance_name", instanceName_.c_str(),
        "resources", resources_.c_str(),
        "logger", &logger_,
        "visible", visible_);
```
-->
```
PyObject* kwargs = Py_BuildValue("{ss,ss,sI,si}",
        "instance_name", instanceName_.c_str(),
        "resources", resources_.c_str(),
        "logger", &logger_,
        "visible", visible_);
```
(The python argument format for the logger pointer should be ` I [unsigned int]` rather than` L = [long long]` to work for both 32bit and 64bit. Alternatively, `n [Py_ssize_t]` also works for both platforms.)

Would be happy if this could enter one the upcoming versions. I am unsure how to proceed from this finding. Somebody able to create an issue from this discussion?

**Reply by markaren** [reply]
Just make an issue with pretty much what you described here, and hopefully follow up with a pull request.

**Reply by havifloja** [reply]
Thanks for your quick reply. I created the issue as https://github.com/NTNU-IHB/PythonFMU/issues/180
